### PR TITLE
Eliminate strings-domain component usages

### PR DESCRIPTION
**Pull request by SamChou19815**
## Summary

~~`strings-domain` is an existing component that does something equivalent to JS's `with`:~~

~~const obj = { foo: 2, bar: 3}~~
~~with (obj) { // destruct obj and make every field in scope~~
~~console.log(foo + bar) // 5 is logged~~
~~}~~

~~Use of `width` is banned by almost all linters by default, so we should also not using components with similar mechanism.~~

The `strings-domain` component also doesn't work well with TS and results in many red squiggly lines. There is currently no way in Vue 2 to make this well typed, because the slots are not part of the type signature of vue components, so there is no way to communicate to the type system that the type of the prop `value` is linked to the type of the named slot `key`.

In this PR, I removed them all and just replaced them with normal field access. As a result, the code should be more readable, and it should be more obvious what kind of data we are using from the json.

## Note

This PR also removes support for linking medium article in the project page, because no project is using it and it probably won't be used later as well.

## Test Plan

Screenshots to show that the website is still working:

Screen Shot 2021-02-26 at 18 50 34
Screen Shot 2021-02-26 at 18 50 37
Screen Shot 2021-02-26 at 18 50 41
Screen Shot 2021-02-26 at 18 50 48
Screen Shot 2021-02-26 at 18 50 52
Screen Shot 2021-02-26 at 18 50 58
Screen Shot 2021-02-26 at 18 51 02
Screen Shot 2021-02-26 at 18 51 05
Screen Shot 2021-02-26 at 18 51 10
Screen Shot 2021-02-26 at 18 51 12
Screen Shot 2021-02-26 at 18 51 16
Screen Shot 2021-02-26 at 18 51 18
Screen Shot 2021-02-26 at 18 51 20
Screen Shot 2021-02-26 at 18 51 25
Screen Shot 2021-02-26 at 18 51 28
Screen Shot 2021-02-26 at 18 51 30

### CODE REVIEW

**Comment by ewlsh** [comment]
I am back from the dead to give some context on this "feature"...

1. It is _not_ like `with`. It offers object destructuring in Vue templates:

```vue
<strings-domain #key="{ a, b, c }" :value="obj" />
```

```js
const { a, b, c } = obj;
```

2. It dates back to when there was only basic support for type-checking in templates. I'm assuming you're using Vetur to check templates. In that case, it _could_ be properly typed - it probably just isn't.

3. It exists primarily to discourage "bad" code which directly accessed property chains:

```vue
<component :text="a.b.c" />
```

This can be dangerous because a "hole" in the chain can cause the template rendering to error out and break the site. The alternative was to define everything in component code _or_ to just use key strings.

We originally had a lib (oh `Strings`) which handled all of this behind-the-scenes using key strings, the goal was to get rid of it and use JavaScript/TypeScript objects directly...

Unfortunately optional chaining wasn't supported in Vue templates (I tried - https://github.com/ewlsh/acorn-optional-chaining) so `strings-domain` was the alternative I wrote.

**Comment by SamChou19815** [comment]
> I am back from the dead to give some context on this "feature"...
> 
> 1. It is _not_ like `with`. It offers object destructuring in Vue templates:
> 
> ```
> <strings-domain #key="{ a, b, c }" :value="obj" />
> ```
> 
> ```js
> const { a, b, c } = obj;
> ```
> 
> 1. It dates back to when there was only basic support for type-checking in templates. I'm assuming you're using Vetur to check templates. In that case, it _could_ be properly typed - it probably just isn't.
> 2. It exists primarily to discourage "bad" code which directly accessed property chains:
> 
> ```
> <component :text="a.b.c" />
> ```
> 
> This can be dangerous because a "hole" in the chain can cause the template rendering to error out and break the site. The alternative was to define everything in component code _or_ to just use key strings.
> 
> We originally had a lib (oh `Strings`) which handled all of this behind-the-scenes using key strings, the goal was to get rid of it and use JavaScript/TypeScript objects directly...
> 
> Unfortunately optional chaining wasn't supported in Vue templates (I tried - https://github.com/ewlsh/acorn-optional-chaining) so `strings-domain` was the alternative I wrote.

@ewlsh 

2.

I don't think it can be typed. It's not like react component where you can specify the exact type of children. Vue's slot isn't even part of the type signature, and I see no path of making vue aware that the key slot and the prop's type is connected.

It _could_ be typed in a distant future where the slot is part of the type signature, but that probably won't happen for years. Meanwhile, while I remove those `strings-domain`, I was able to immediately find some hidden type errors and fix them. Therefore, it's still beneficial to not rely on it.



3.

I don't think using `strings-domain` or similar solution is a solution to avoid bad code of creating "holes". Those holes are undefined values, but it should be the job of TS to catch those errors instead of just programming defensively on your own. That's why we turn on the `strict` flag in TS config because it helps to identify null safety issues.

From my observation, `strings-domain` also doesn't quite help solving the problem either. To prevent crashing due to undefined field, the code is already using a lot of `v-if` to do that.

Finally, I believe that crashing the page might not be a bad idea. We are already using a static site generator, which means it will evaluate all the JS. If things will crash when user visits it, it will also crash while building. We will catch the error, with or without typescript.

**Comment by ewlsh** [comment]
@SamChou19815 

> I don't think it can be typed. It's not like react component where you can specify the exact type of children. Vue's slot isn't even part of the type signature, and I see no path of making vue aware that the key slot and the prop's type is connected.
> 
> It could be typed in a distant future where the slot is part of the type signature, but that probably won't happen for years. Meanwhile, while I remove those strings-domain, I was able to immediately find some hidden type errors and fix them. Therefore, it's still beneficial to not rely on it.

Agreed (in Vue 2)

> I don't think using strings-domain or similar solution is a solution to avoid bad code of creating "holes". Those holes are undefined values, but it should be the job of TS to catch those errors instead of just programming defensively on your own. That's why we turn on the strict flag in TS config because it helps to identify null safety issues.
> 
> From my observation, strings-domain also doesn't quite help solving the problem either. To prevent crashing due to undefined field, the code is already using a lot of v-if to do that.
> 
> Finally, I believe that crashing the page might not be a bad idea. We are already using a static site generator, which means it will evaluate all the JS. If things will crash when user visits it, it will also crash while building. We will catch the error, with or without typescript.

Yes, `strings-domain` was born out of a time when TypeScript wasn't standard in this codebase and is a pattern carried over from the `Strings` days. TypeScript can't "catch" holes in our JSON files alone, but if the JSON imports are typed defensively you can create a decent pattern.

**I'm not opposed to getting rid of it**, I was just clarifying its context - hence `"feature"` (heavy air quotes) :grimacing: 

I would note that Vue 3 has much better TypeScript support and its composition API is :partying_face: so once the Vue Jamstack finishes migrating, I think this ugliness will be moot and we can do optional chaining, slot typing, and other fun things.

**Comment by ewlsh** [comment]
@SamChou19815 document in the commit message that this also removes Medium support for project pages and you've got my approval.

I believe CUReviews used to use that but no one appears to anymore.

File: packages/website/src/components/ProjectLearnMore.vue [hunk 1]
```diff
@@ -1,70 +1,49 @@
...
+    v-if="
+      enableAll ||
+        (!website && !playstore && !appstore) ||
+        ((playstore || appstore || website) && (ios_github || android_github || github))
+    "
   >
-    <page-section
-      v-if="
-        enableAll ||
-          (!website && !playstore && !appstore && medium) ||
```
Comment (ewlsh) [inline packages/website/src/components/ProjectLearnMore.vue:19]: Document the `medium` support removal in the commit please. :smile_cat:

**Review by ewlsh — CHANGES_REQUESTED** [review]

File: packages/website/src/components/ProjectLearnMore.vue (hunk 1 above)
Comment (SamChou19815) [reply packages/website/src/components/ProjectLearnMore.vue:19]: @ewlsh  Documented in PR comment and commit message

**Review by ewlsh — APPROVED** [review]

**Pull request merged**